File: Python/OneLang/One/Transforms/InferTypesPlugins/BasicTypeInfer.py
from OneLangStdLib import *
import OneLang.One.Transforms.InferTypesPlugins.Helpers.InferTypesPlugin as inferTypesPlug
import OneLang.One.Ast.Expressions as exprs
import OneLang.One.Ast.References as refs
import OneLang.One.Ast.AstTypes as astTypes

class BasicTypeInfer(inferTypesPlug.InferTypesPlugin):
    def __init__(self):
        super().__init__("BasicTypeInfer")
    
    def can_detect_type(self, expr):
        return True
    
    def detect_type(self, expr):
        lit_types = self.main.current_file.literal_types
        
        if isinstance(expr, exprs.CastExpression):
            expr.set_actual_type(expr.new_type)
        elif isinstance(expr, exprs.ParenthesizedExpression):
            expr.set_actual_type(expr.expression.get_type())
        elif isinstance(expr, refs.ThisReference):
            expr.set_actual_type(expr.cls_.type, False, False)
        elif isinstance(expr, refs.SuperReference):
            expr.set_actual_type(expr.cls_.type, False, False)
        elif isinstance(expr, refs.MethodParameterReference):
            expr.set_actual_type(expr.decl.type, False, False)
        elif isinstance(expr, exprs.BooleanLiteral):
            expr.set_actual_type(lit_types.boolean)
        elif isinstance(expr, exprs.NumericLiteral):
            expr.set_actual_type(lit_types.numeric)
        elif isinstance(expr, exprs.StringLiteral) or isinstance(expr, exprs.TemplateString):
            expr.set_actual_type(lit_types.string)
        elif isinstance(expr, exprs.RegexLiteral):
            expr.set_actual_type(lit_types.regex)
        elif isinstance(expr, exprs.InstanceOfExpression):
            expr.set_actual_type(lit_types.boolean)
        elif isinstance(expr, exprs.NullLiteral):
            expr.set_actual_type(expr.expected_type if expr.expected_type != None else astTypes.NullType.instance)
        elif isinstance(expr, refs.VariableDeclarationReference):
            expr.set_actual_type(expr.decl.type)
        elif isinstance(expr, refs.ForeachVariableReference):
            expr.set_actual_type(expr.decl.type)
        elif isinstance(expr, refs.ForVariableReference):
            expr.set_actual_type(expr.decl.type)
        elif isinstance(expr, refs.CatchVariableReference):
            expr.set_actual_type(expr.decl.type or self.main.current_file.literal_types.error)
        elif isinstance(expr, exprs.UnaryExpression):
            operand_type = expr.operand.get_type()
            if isinstance(operand_type, astTypes.ClassType):
                op_id = f'''{expr.operator}{operand_type.decl.name}'''
                
                if op_id == "-TsNumber":
                    expr.set_actual_type(lit_types.numeric)
                elif op_id == "+TsNumber":
                    expr.set_actual_type(lit_types.numeric)
                elif op_id == "!TsBoolean":
                    expr.set_actual_type(lit_types.boolean)
                elif op_id == "++TsNumber":
                    expr.set_actual_type(lit_types.numeric)
                elif op_id == "--TsNumber":
                    expr.set_actual_type(lit_types.numeric)
                else:
                    pass
            elif isinstance(operand_type, astTypes.AnyType):
                expr.set_actual_type(astTypes.AnyType.instance)
            else:
                pass
        elif isinstance(expr, exprs.BinaryExpression):
            left_type = expr.left.get_type()
            right_type = expr.right.get_type()
            is_eq_or_neq = expr.operator == "==" or expr.operator == "!="
            if expr.operator == "=":
                if astTypes.TypeHelper.is_assignable_to(right_type, left_type):
                    expr.set_actual_type(left_type, False, True)
                else:
                    raise Error(f'''Right-side expression ({right_type.repr()}) is not assignable to left-side ({left_type.repr()}).''')
            elif is_eq_or_neq:
                expr.set_actual_type(lit_types.boolean)
            elif isinstance(left_type, astTypes.ClassType) and isinstance(right_type, astTypes.ClassType):
                if left_type.decl == lit_types.numeric.decl and right_type.decl == lit_types.numeric.decl and expr.operator in ["-", "+", "-=", "+=", "%", "/"]:
                    expr.set_actual_type(lit_types.numeric)
                elif left_type.decl == lit_types.numeric.decl and right_type.decl == lit_types.numeric.decl and expr.operator in ["<", "<=", ">", ">="]:
                    expr.set_actual_type(lit_types.boolean)
                elif left_type.decl == lit_types.string.decl and right_type.decl == lit_types.string.decl and expr.operator in ["+", "+="]:
                    expr.set_actual_type(lit_types.string)
                elif left_type.decl == lit_types.string.decl and right_type.decl == lit_types.string.decl and expr.operator in ["<="]:
                    expr.set_actual_type(lit_types.boolean)
                elif left_type.decl == lit_types.boolean.decl and right_type.decl == lit_types.boolean.decl and expr.operator in ["||", "&&"]:
                    expr.set_actual_type(lit_types.boolean)
                elif left_type.decl == lit_types.string.decl and right_type.decl == lit_types.map.decl and expr.operator == "in":
                    expr.set_actual_type(lit_types.boolean)
                else:
                    pass
            elif isinstance(left_type, astTypes.EnumType) and isinstance(right_type, astTypes.EnumType):
                if left_type.decl == right_type.decl and is_eq_or_neq:
                    expr.set_actual_type(lit_types.boolean)
                else:
                    pass
            elif isinstance(left_type, astTypes.AnyType) and isinstance(right_type, astTypes.AnyType):
                expr.set_actual_type(astTypes.AnyType.instance)
            else:
                pass
        elif isinstance(expr, exprs.ConditionalExpression):
            true_type = expr.when_true.get_type()
            false_type = expr.when_false.get_type()
            if expr.expected_type != None:
                if not astTypes.TypeHelper.is_assignable_to(true_type, expr.expected_type):
                    raise Error(f'''Conditional expression expects {expr.expected_type.repr()} but got {true_type.repr()} as true branch''')
                if not astTypes.TypeHelper.is_assignable_to(false_type, expr.expected_type):
                    raise Error(f'''Conditional expression expects {expr.expected_type.repr()} but got {false_type.repr()} as false branch''')
                expr.set_actual_type(expr.expected_type)
            elif astTypes.TypeHelper.is_assignable_to(true_type, false_type):
                expr.set_actual_type(false_type)
            elif astTypes.TypeHelper.is_assignable_to(false_type, true_type):
                expr.set_actual_type(true_type)
            else:
                raise Error(f'''Different types in the whenTrue ({true_type.repr()}) and whenFalse ({false_type.repr()}) expressions of a conditional expression''')
        elif isinstance(expr, exprs.NullCoalesceExpression):
            default_type = expr.default_expr.get_type()
            if_null_type = expr.expr_if_null.get_type()
            if not astTypes.TypeHelper.is_assignable_to(if_null_type, default_type):
                self.error_man.throw(f'''Null-coalescing operator tried to assign incompatible type "{if_null_type.repr()}" to "{default_type.repr()}"''')
            else:
                expr.set_actual_type(default_type)
        elif isinstance(expr, exprs.AwaitExpression):
            expr_type = expr.expr.get_type()
            if isinstance(expr_type, astTypes.ClassType) and expr_type.decl == lit_types.promise.decl:
                expr.set_actual_type((expr_type).type_arguments[0], True)
            else:
                self.error_man.throw(f'''Expected promise type ({lit_types.promise.repr()}) for await expression, but got {expr_type.repr()}''')
        else:
            return False
        
        return True

File: PHP/src/OneLang/One/Transforms/InferTypesPlugins/BasicTypeInfer.php
<?php

namespace OneLang\One\Transforms\InferTypesPlugins\BasicTypeInfer;

use OneLang\One\Transforms\InferTypesPlugins\Helpers\InferTypesPlugin\InferTypesPlugin;
use OneLang\One\Ast\Expressions\Expression;
use OneLang\One\Ast\Expressions\CastExpression;
use OneLang\One\Ast\Expressions\ParenthesizedExpression;
use OneLang\One\Ast\Expressions\BooleanLiteral;
use OneLang\One\Ast\Expressions\NumericLiteral;
use OneLang\One\Ast\Expressions\StringLiteral;
use OneLang\One\Ast\Expressions\TemplateString;
use OneLang\One\Ast\Expressions\RegexLiteral;
use OneLang\One\Ast\Expressions\InstanceOfExpression;
use OneLang\One\Ast\Expressions\NullLiteral;
use OneLang\One\Ast\Expressions\UnaryExpression;
use OneLang\One\Ast\Expressions\BinaryExpression;
use OneLang\One\Ast\Expressions\ConditionalExpression;
use OneLang\One\Ast\Expressions\NewExpression;
use OneLang\One\Ast\Expressions\NullCoalesceExpression;
use OneLang\One\Ast\Expressions\LambdaCallExpression;
use OneLang\One\Ast\Expressions\AwaitExpression;
use OneLang\One\Ast\References\ThisReference;
use OneLang\One\Ast\References\MethodParameterReference;
use OneLang\One\Ast\References\VariableDeclarationReference;
use OneLang\One\Ast\References\ForeachVariableReference;
use OneLang\One\Ast\References\ForVariableReference;
use OneLang\One\Ast\References\SuperReference;
use OneLang\One\Ast\References\CatchVariableReference;
use OneLang\One\Ast\AstTypes\ClassType;
use OneLang\One\Ast\AstTypes\AnyType;
use OneLang\One\Ast\AstTypes\EnumType;
use OneLang\One\Ast\AstTypes\NullType;
use OneLang\One\Ast\AstTypes\TypeHelper;

class BasicTypeInfer extends InferTypesPlugin {
    function __construct() {
        parent::__construct("BasicTypeInfer");
        
    }
    
    function canDetectType($expr) {
        return true;
    }
    
    function detectType($expr) {
        $litTypes = $this->main->currentFile->literalTypes;
        
        if ($expr instanceof CastExpression)
            $expr->setActualType($expr->newType);
        else if ($expr instanceof ParenthesizedExpression)
            $expr->setActualType($expr->expression->getType());
        else if ($expr instanceof ThisReference)
            $expr->setActualType($expr->cls->type, false, false);
        else if ($expr instanceof SuperReference)
            $expr->setActualType($expr->cls->type, false, false);
        else if ($expr instanceof MethodParameterReference)
            $expr->setActualType($expr->decl->type, false, false);
        else if ($expr instanceof BooleanLiteral)
            $expr->setActualType($litTypes->boolean);
        else if ($expr instanceof NumericLiteral)
            $expr->setActualType($litTypes->numeric);
        else if ($expr instanceof StringLiteral || $expr instanceof TemplateString)
            $expr->setActualType($litTypes->string);
        else if ($expr instanceof RegexLiteral)
            $expr->setActualType($litTypes->regex);
        else if ($expr instanceof InstanceOfExpression)
            $expr->setActualType($litTypes->boolean);
        else if ($expr instanceof NullLiteral)
            $expr->setActualType($expr->expectedType !== null ? $expr->expectedType : NullType::$instance);
        else if ($expr instanceof VariableDeclarationReference)
            $expr->setActualType($expr->decl->type);
        else if ($expr instanceof ForeachVariableReference)
            $expr->setActualType($expr->decl->type);
        else if ($expr instanceof ForVariableReference)
            $expr->setActualType($expr->decl->type);
        else if ($expr instanceof CatchVariableReference)
            $expr->setActualType($expr->decl->type ?? $this->main->currentFile->literalTypes->error);
        else if ($expr instanceof UnaryExpression) {
            $operandType = $expr->operand->getType();
            if ($operandType instanceof ClassType) {
                $opId = $expr->operator . $operandType->decl->name;
                
                if ($opId === "-TsNumber")
                    $expr->setActualType($litTypes->numeric);
                else if ($opId === "+TsNumber")
                    $expr->setActualType($litTypes->numeric);
                else if ($opId === "!TsBoolean")
                    $expr->setActualType($litTypes->boolean);
                else if ($opId === "++TsNumber")
                    $expr->setActualType($litTypes->numeric);
                else if ($opId === "--TsNumber")
                    $expr->setActualType($litTypes->numeric);
                else { }
            }
            else if ($operandType instanceof AnyType)
                $expr->setActualType(AnyType::$instance);
            else { }
        }
        else if ($expr instanceof BinaryExpression) {
            $leftType = $expr->left->getType();
            $rightType = $expr->right->getType();
            $isEqOrNeq = $expr->operator === "==" || $expr->operator === "!=";
            if ($expr->operator === "=") {
                if (TypeHelper::isAssignableTo($rightType, $leftType))
                    $expr->setActualType($leftType, false, true);
                else
                    throw new \OneLang\Core\Error("Right-side expression (" . $rightType->repr() . ") is not assignable to left-side (" . $leftType->repr() . ").");
            }
            else if ($isEqOrNeq)
                $expr->setActualType($litTypes->boolean);
            else if ($leftType instanceof ClassType && $rightType instanceof ClassType) {
                if ($leftType->decl === $litTypes->numeric->decl && $rightType->decl === $litTypes->numeric->decl && in_array($expr->operator, array("-", "+", "-=", "+=", "%", "/")))
                    $expr->setActualType($litTypes->numeric);
                else if ($leftType->decl === $litTypes->numeric->decl && $rightType->decl === $litTypes->numeric->decl && in_array($expr->operator, array("<", "<=", ">", ">=")))
                    $expr->setActualType($litTypes->boolean);
                else if ($leftType->decl === $litTypes->string->decl && $rightType->decl === $litTypes->string->decl && in_array($expr->operator, array("+", "+=")))
                    $expr->setActualType($litTypes->string);
                else if ($leftType->decl === $litTypes->string->decl && $rightType->decl === $litTypes->string->decl && in_array($expr->operator, array("<=")))
                    $expr->setActualType($litTypes->boolean);
                else if ($leftType->decl === $litTypes->boolean->decl && $rightType->decl === $litTypes->boolean->decl && in_array($expr->operator, array("||", "&&")))
                    $expr->setActualType($litTypes->boolean);
                else if ($leftType->decl === $litTypes->string->decl && $rightType->decl === $litTypes->map->decl && $expr->operator === "in")
                    $expr->setActualType($litTypes->boolean);
                else { }
            }
            else if ($leftType instanceof EnumType && $rightType instanceof EnumType) {
                if ($leftType->decl === $rightType->decl && $isEqOrNeq)
                    $expr->setActualType($litTypes->boolean);
                else { }
            }
            else if ($leftType instanceof AnyType && $rightType instanceof AnyType)
                $expr->setActualType(AnyType::$instance);
            else { }
        }
        else if ($expr instanceof ConditionalExpression) {
            $trueType = $expr->whenTrue->getType();
            $falseType = $expr->whenFalse->getType();
            if ($expr->expectedType !== null) {
                if (!TypeHelper::isAssignableTo($trueType, $expr->expectedType))
                    throw new \OneLang\Core\Error("Conditional expression expects " . $expr->expectedType->repr() . " but got " . $trueType->repr() . " as true branch");
                if (!TypeHelper::isAssignableTo($falseType, $expr->expectedType))
                    throw new \OneLang\Core\Error("Conditional expression expects " . $expr->expectedType->repr() . " but got " . $falseType->repr() . " as false branch");
                $expr->setActualType($expr->expectedType);
            }
            else if (TypeHelper::isAssignableTo($trueType, $falseType))
                $expr->setActualType($falseType);
            else if (TypeHelper::isAssignableTo($falseType, $trueType))
                $expr->setActualType($trueType);
            else
                throw new \OneLang\Core\Error("Different types in the whenTrue (" . $trueType->repr() . ") and whenFalse (" . $falseType->repr() . ") expressions of a conditional expression");
        }
        else if ($expr instanceof NullCoalesceExpression) {
            $defaultType = $expr->defaultExpr->getType();
            $ifNullType = $expr->exprIfNull->getType();
            if (!TypeHelper::isAssignableTo($ifNullType, $defaultType))
                $this->errorMan->throw("Null-coalescing operator tried to assign incompatible type \"" . $ifNullType->repr() . "\" to \"" . $defaultType->repr() . "\"");
            else
                $expr->setActualType($defaultType);
        }
        else if ($expr instanceof AwaitExpression) {
            $exprType = $expr->expr->getType();
            if ($exprType instanceof ClassType && $exprType->decl === $litTypes->promise->decl)
                $expr->setActualType(($exprType)->typeArguments[0], true);
            else
                $this->errorMan->throw("Expected promise type (" . $litTypes->promise->repr() . ") for await expression, but got " . $exprType->repr());
        }
        else
            return false;
        
        return true;
    }
}
